### ISSUE TITLE
Atom doesn't save the file tree upon closing the program

### ISSUE THREAD

**Issue by RyanIMinnick** [atom/atom#14315]
### Prerequisites

* [ x  ] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

Atom doesn't save the file tree upon closing the program. The ```application.json``` file clears every time, even with **"Restore Previous Windows on Start"** enabled. I have tried completely wiping and re-installing Atom, choosing a different directory, opening it in safe mode, none of those things have solved the issue.

### Steps to Reproduce

1. Open Atom
2. Open a repository/directory
3. Close Atom
4. Reboot
5. Yield result of a blank editor **insert sad face**

**Expected behavior:** After closing atom, it **should** save my previous project file tree. It doesn't.

**Actual behavior:** Atom opens with a completely blank editor.

**Reproduces how often:** 100% of the time.

### Versions

```
Atom    : 1.16.0
Electron: 1.3.13
Chrome  : 52.0.2743.82
Node    : 6.5.0
```
### Additional Information

I am on the latest version of macOS.

**Comment by rsese** [comment]
Thanks for opening a new issue @RyanIMinnick - cross-referencing your message board post with a lot of the debugging steps you've already tried:

https://discuss.atom.io/t/atom-wont-save-my-default-file-tree/42077

**Comment by RyanIMinnick** [comment]
@rsese Yes, sorry, I should have included that.

**Comment by rsese** [comment]
No problem :v:

Can you confirm if you have `Settings > Core > Restore Previous Windows on Start` checked?

**Comment by RyanIMinnick** [comment]
Yes. I tried it with that turned on, turned off, with "open empty editor" on and off, and now after reinstalling atom restore previous window on start is enabled.

**Comment by rsese** [comment]
> 3. Close Atom

Can you confirm how you close Atom?  Are you clicking the red x in the window title bar or are you quitting the program (e.g. with <kbd>cmd-q</kbd> or `Atom > Quit` in the menu)?

**Comment by RyanIMinnick** [comment]
All return the same result @rsese, my file tree cleared. I've tried all that you've mentioned.

**Comment by rsese** [comment]
Hmm, would you be able to record a screencast or GIF showing these steps?

1. With an application.json file that has no paths, run `cat ~/.atom/storage/application.json`
1. In your project directory start Atom with `atom .`
1. run `cat ~/.atom/storage/application.json`
1. Close Atom from the menu with `Atom > Quit`
1. run `cat ~/.atom/storage/application.json`

So if you could basically match my GIF, that would be great:

![app json-saved](https://cloud.githubusercontent.com/assets/734194/25549758/67fd0f4c-2c39-11e7-82e2-da7080338bd5.gif)

Sorry for the repetition, I know you've gone through a lot of this already based on the message board post but I just want to step back a bit and make sure we're on the same page.

**Comment by rsese** [comment]
Missed the last step of starting Atom again with just `atom`, added the step above.

**Comment by RyanIMinnick** [comment]
I'll give it a try.

**Comment by RyanIMinnick** [comment]
@rsese https://vid.me/8GwF

**Comment by rsese** [comment]
Great thanks @RyanIMinnick I think I see what's happening - it looks like everything works fine when you get to the point where you`Atom > Quit` and then start Atom again.  The project you had open is opened again when you restart and the project path in your `application.json` file is correct to match. 

The issue is at around where there's ~22 seconds left in the video, it looks like you close the Atom window with the red x in the window bar rather than quitting Atom (let me know if that's not the case).  When you close the window, Atom is still running but now there's no project to remember, it's remembering this empty state (as reflected in your `application.json` file).  Max describes this behavior here:

https://github.com/atom/atom/issues/13895#issuecomment-283726843

So basically, if you want Atom to remember the project you had opened, you should quit Atom (with `Atom > Quit` or <kbd>cmd-q</kbd>) rather than just closing the last window.  As Max also mentions, this is macOS specific since on Windows and Linux, if you close the last open Atom window, that quits the app.

So I can reproduce the behavior you're reporting if I close the Atom window rather than quitting Atom, but that's expected behavior.  Let me know if that makes sense of if you have questions @RyanIMinnick.

**Comment by RyanIMinnick** [comment]
Okay... it saves the previous session if I close with CMD+Q or Atom > Quit (Both in the hotbar and navbar)... I wonder why it resets on a soft close but on a hard close it doesn't? Weird behavior. Good to know its normal though.

**Comment by rsese** [comment]
> it saves the previous session if I close with CMD+Q or Atom > Quit (Both in the hotbar and navbar)...

Yep that's right :+1:

> I wonder why it resets on a soft close but on a hard close it doesn't?

It's because when you soft close so there's no windows but Atom is still open, it's remembering that empty state.  When you quit while a project is open, the last state Atom remembers is with that last open project.

> Weird behavior. Good to know its normal though.

Yep I'm glad we figured out what was going on, thanks for your patience!
